File: solution/931. Minimum Falling Path Sum/931_02.py
class Solution:
    def minFallingPathSum(self, matrix: List[List[int]]) -> int:
        n = len(matrix[0])
        
        for i in range(1, n):
            for j in range(n):
                val = matrix[i][j]
                matrix[i][j] = inf
                for k in range(max(0, j - 1), min(n - 1, j + 1) + 1):
                    matrix[i][j] = min(matrix[i][j], matrix[i - 1][k] + val)
        
        return min(matrix[-1])
    

File: solution/931. Minimum Falling Path Sum/931_02.cpp
class Solution {
public:
    int minFallingPathSum(vector<vector<int>>& matrix) {
        int n = matrix.size();
        
        for (int i = 1; i < n; ++i) {
            for (int j = 0; j < n; ++j) {
                int val = matrix[i][j];
                matrix[i][j] = INT_MAX;
                for (int k = max(0, j - 1); k <= min(n - 1, j + 1); ++k)
                    matrix[i][j] = min(matrix[i][j], matrix[i - 1][k] + val);
            }
        }
        return *min_element(matrix.back().begin(), matrix.back().end());
    }
};
